internal/pkg/discovery/discovery.go

```diff
@@ -0,0 +1,147 @@
+package discovery
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"regexp"
+	"slices"
+	"strings"
+
+	"go.uber.org/multierr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+)
+
+type DiscoveryOptions struct {
+	BatchSize int64
+	LogWriter io.Writer
+
+	// MustExistResources is a list of resources that must exist in the cluster.
+	// This can be used as a sanity check to ensure that the discovery process is working as expected.
+	// If a resource does not exist, the discovery process will fail.
+	// If the list is empty, no resources are required to exist.
+	MustExistResources []string
+
+	// IgnoreResources is a list of resources to ignore during discovery.
+	IgnoreResources []*regexp.Regexp
+}
+
+// GetBatchSize returns the set batch size for listing objects or the default.
+func (opts DiscoveryOptions) GetBatchSize() int64 {
+	if opts.BatchSize == 0 {
+		return 500
+	}
+	return opts.BatchSize
+}
+
+// GetLogWriter returns the set batch size for listing objects or io.Discard as default.
+func (opts DiscoveryOptions) GetLogWriter() io.Writer {
+	if opts.LogWriter == nil {
+		return io.Discard
+	}
+	return opts.LogWriter
+}
+
+// DiscoverObjects discovers all objects in the cluster and calls the provided callback for each list of objects.
+// The callback can be called multiple times with the same
```

Comment:
Incomplete comment, should this be

```suggestion
// The callback can be called multiple times with the same object
```

Reply:
```
The callback can be called multiple times with the same object kind. Objects are unique in general, but might not be if the aggregate API-Server does not implement batching correctly.
```

Reply:
```
// DiscoverObjects discovers all objects in the cluster and calls the provided callback for each list of objects.
// The callback can be called multiple times with the same object kind. Objects are unique in general,
// but the callback should be able to handle duplicates. Some API-Servers do not implement list batching correctly and thus might introduce duplicates.
```